homeassistant/components/pax_ble/calima.py

```diff
@@ -0,0 +1,323 @@
+from bleak.exc import BleakError
+import asyncio
+import binascii
+import math
+from collections import namedtuple
+import logging
+
+_LOGGER = logging.getLogger(__name__)
+
+Fanspeeds = namedtuple('Fanspeeds', 'Humidity Light Trickle')
+Fanspeeds.__new__.__defaults__ = (2250, 1625, 1000)
+Time = namedtuple('Time', 'DayOfWeek Hour Minute Second')
+Sensitivity = namedtuple('Sensitivity', 'HumidityOn Humidity LightOn Light')
+LightSensorSettings = namedtuple('LightSensorSettings', 'DelayedStart RunningTime')
+HeatDistributorSettings = namedtuple('HeatDistributorSettings', 'TemperatureLimit FanSpeedBelow FanSpeedAbove')
+SilentHours = namedtuple('SilentHours', 'On StartingHour StartingMinute EndingHour EndingMinute')
+TrickleDays = namedtuple('TrickleDays', 'Weekdays Weekends')
+BoostMode = namedtuple('BoostMode', 'OnOff Speed Seconds')
+
+FanState = namedtuple('FanState', 'Humidity Temp Light RPM Mode')
+
+# Stolen defines for each characteristic (taken from a decompiled Android App)
+CHARACTERISTIC_APPEARANCE = "00002a01-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_AUTOMATIC_CYCLES = "f508408a-508b-41c6-aa57-61d1fd0d5c39"
+CHARACTERISTIC_BASIC_VENTILATION = "faa49e09-a79c-4725-b197-bdc57c67dc32"
+CHARACTERISTIC_BOOST = "118c949c-28c8-4139-b0b3-36657fd055a9"
+CHARACTERISTIC_CLOCK = "6dec478e-ae0b-4186-9d82-13dda03c0682"
+CHARACTERISTIC_DEVICE_NAME = "00002a00-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_FACTORY_SETTINGS_CHANGED = "63b04af9-24c0-4e5d-a69c-94eb9c5707b4"
+CHARACTERISTIC_FAN_DESCRIPTION = "b85fa07a-9382-4838-871c-81d045dcc2ff"
+CHARACTERISTIC_FIRMWARE_REVISION = "00002a26-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_HARDWARE_REVISION = "00002a27-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_SOFTWARE_REVISION = "00002a28-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_LED = "8b850c04-dc18-44d2-9501-7662d65ba36e"
+CHARACTERISTIC_LEVEL_OF_FAN_SPEED = "1488a757-35bc-4ec8-9a6b-9ecf1502778e"
+CHARACTERISTIC_MANUFACTURER_NAME = "00002a29-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_MODE = "90cabcd1-bcda-4167-85d8-16dcd8ab6a6b"
+CHARACTERISTIC_MODEL_NAME = "00002a00-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_MODEL_NUMBER = "00002a24-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_NIGHT_MODE = "b5836b55-57bd-433e-8480-46e4993c5ac0"
+CHARACTERISTIC_PIN_CODE = "4cad343a-209a-40b7-b911-4d9b3df569b2"
+CHARACTERISTIC_PIN_CONFIRMATION = "d1ae6b70-ee12-4f6d-b166-d2063dcaffe1"
+CHARACTERISTIC_RESET = "ff5f7c4f-2606-4c69-b360-15aaea58ad5f"
+CHARACTERISTIC_SENSITIVITY = "e782e131-6ce1-4191-a8db-f4304d7610f1"
+CHARACTERISTIC_SENSOR_DATA = "528b80e8-c47a-4c0a-bdf1-916a7748f412"
+CHARACTERISTIC_SERIAL_NUMBER = "00002a25-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_SOFTWARE_REVISION = "00002a28-0000-1000-8000-00805f9b34fb"
+CHARACTERISTIC_STATUS = "25a824ad-3021-4de9-9f2f-60cf8d17bded"
+CHARACTERISTIC_TEMP_HEAT_DISTRIBUTOR = "a22eae12-dba8-49f3-9c69-1721dcff1d96"
+CHARACTERISTIC_TIME_FUNCTIONS = "49c616de-02b1-4b67-b237-90f66793a6f2"
+
+class Calima:
```

Comment:
This should be moved to a separate library published on PyPi.

Also, the bleak library is very prone to generate errors such as timeouts and connection failures because it only tries operations a single time. To improve the user experience, you should really consider using the helper `bleak_retry_connector` library by @bdraco.

Here are a couple of examples of how it can be used:
- https://github.com/Bluetooth-Devices/led-ble/tree/main/src/led_ble
- https://github.com/emontnemery/py-dormakaba-dkey/tree/main/py_dormakaba_dkey

You can contact me on discord if you want: emontnemery#6618